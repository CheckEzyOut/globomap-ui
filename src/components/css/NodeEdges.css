/*
Copyright 2017 Globo.com

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.sub-node-edges .edges-in,
.sub-node-edges .edges-out {
  position: relative;
}

.sub-node-edges .edges-btn {
  box-sizing: content-box;
  width: 16px;
  height: 13px;
  font-size: 12px;
  color: white;
  padding: 2px;
  border: 0;
  border-radius: 2px;
  text-align: center;
  cursor: pointer;
}

.sub-node-edges .edges-btn:hover {
  opacity: 0.9;
}

.sub-node-edges .edges-content {
  position: absolute;
  top: 35px;
  min-width: 100px;
  margin-bottom: 20px;
  border-radius: 2px;
  text-align: left;
  background-color: white;
  box-shadow: 0 2px 15px rgba(0, 0, 0, 0.3);
  z-index: 9;
  cursor: default;
}

.sub-node-edges .edges-content .edge-item {
  display: block;
  font-size: 13px;
  white-space: nowrap;
  color: #333;
  border-top: 1px solid #ddd;
}

.sub-node-edges .edges-content .edge-item:last-child {
  border-bottom: 0;
}

.sub-node-edges .edge-item .item-properties {
  display: none;
  background-color: #dfe2e2;
  padding: 4px 0;
  box-shadow: inset 1px 1px 2px rgba(0, 0, 0, 0.2);
}

.sub-node-edges .edge-item.open .item-properties {
  display: block;
}

.sub-node-edges .edge-item .icon-right { display: inline; }
.sub-node-edges .edge-item.open .icon-right { display: none; }

.sub-node-edges .edge-item .icon-down { display: none; }
.sub-node-edges .edge-item.open .icon-down { display: inline; }

.sub-node-edges .edge-item.no-prop .icon-right {
  color: #bbb;
}

.sub-node-edges .edge-item .edge-item-prop {
  display: inline-block;
  padding: 4px 10px 6px;
  cursor: pointer;
}

.sub-node-edges .edge-item.no-prop .edge-item-prop {
  cursor: default;
}

.sub-node-edges .edges-content .v-line {
  position: absolute;
  top: 0;
  bottom: 0;
  width: 3px;
}

.sub-node-edges.left .edges-content { left: 0; }
.sub-node-edges.left .edges-content .v-line {
  left: -3px;
  border-radius: 3px 0 0 3px;
}

.sub-node-edges.right .edges-content { right: 0; }
.sub-node-edges.right .edges-content .v-line {
  right: -3px;
  border-radius: 0 3px 3px 0;
}

.sub-node-edges .tooltip-arrow {
  position: absolute;
  top: -8px;
  width: 0;
  height: 0;
  border-left: 8px solid transparent;
  border-right: 8px solid transparent;
  border-bottom: 8px solid white;
}

.sub-node-edges.left .tooltip-arrow {
  left: 4px;
}

.sub-node-edges.right .tooltip-arrow {
  right: 4px;
}

.sub-node-edges .edges-content-head {
  text-align: left;
  font-size: 16px;
  padding: 10px;
}

.sub-node-edges .close-tooltip-btn {
  position: absolute;
  top: 8px;
  right: 6px;
  border: 0;
  background-color: transparent;
  font-size: 14px;
  color: #aaa;
}

.sub-node-edges .close-tooltip-btn:hover {
  color: #333;
}

