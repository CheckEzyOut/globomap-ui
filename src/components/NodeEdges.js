import React, { Component } from 'react';
import Properties from './Properties';
import './css/NodeEdges.css';

class NodeEdges extends Component {

  constructor(props) {
    super(props);
    this.state = {
      inOpen: false,
      outOppen: false
    }

    this.buildEdgeItems = this.buildEdgeItems.bind(this);
    this.onOpenIn = this.onOpenIn.bind(this);
    this.onOpenOut = this.onOpenOut.bind(this);
    this.onOpenProp = this.onOpenProp.bind(this);
    this.closeAll = this.closeAll.bind(this);
    this.handleOutsideClick = this.handleOutsideClick.bind(this);
  }

  render() {
    let edges = this.props.edges;
    if(edges === undefined) {
      return <div className="sub-node-edges"></div>;
    }

    let colorCls = this.props.graphs.filter((graph) => {
        return graph.name === edges.graph;
    })[0].colorClass;

    let edgesInOut = [
      { dir: 'in', toggleFn: this.onOpenIn, openState: this.state.inOpen, items: this.buildEdgeItems(edges.in) },
      { dir: 'out', toggleFn: this.onOpenOut, openState: this.state.outOpen, items: this.buildEdgeItems(edges.out) }
    ].map((elem) => {
      return elem.items.length > 0
              ? (<div key={elem.dir} className={'edges-' + elem.dir}>
                  <button className={'edges-btn ' + colorCls} onClick={elem.toggleFn}>
                    <i className={'fa fa-arrow-'+ (elem.dir === 'in' ? 'right' : 'left')}></i>
                  </button>
                  {elem.openState &&
                    <div className="edges-content" onClick={e => e.stopPropagation()}>
                      <div className="tooltip-arrow"></div>
                      <div className={'v-line ' + colorCls}></div>
                      <div className="edges-content-head">
                        Links
                        <button className="close-tooltip-btn" onClick={elem.toggleFn}>
                          <i className="fa fa-close"></i>
                        </button>
                      </div>
                      {elem.items}
                    </div>}
                  </div>)
              : null;
    });

    return <div className={'sub-node-edges ' + this.props.position}
                ref={node => { this.node = node; }}>
             {edgesInOut}
           </div>;
  }

  buildEdgeItems(edges) {
    return edges.map((edge, i) => {
      let noProp = edge.properties.length === 0 ? ' no-prop' : '';

      return <span key={i} className={'edge-item' + noProp} title={edge.name}>
               <span className="edge-item-prop" onClick={(e) => this.onOpenProp(e, edge)}>
                 <i className="icon-right fa fa-caret-right"></i>
                 <i className="icon-down fa fa-caret-down"></i>
                 &nbsp;<strong>{edge.type}</strong>: {edge.name}
               </span>
               <Properties properties={edge.properties} />
             </span>;
    });
  }

  handleOutsideClick(e) {
    if (this.node && this.node.contains(e.target)) {
      return;
    }
    this.closeAll();
  }

  closeAll() {
    this.setState({ inOpen: false, outOpen: false });
  }

  onOpenIn(event) {
    event.stopPropagation();
    this.setState({ inOpen: !this.state.inOpen });
  }

  onOpenOut(event) {
    event.stopPropagation();
    this.setState({ outOpen: !this.state.outOpen });
  }

  onOpenProp(event, edge) {
    event.stopPropagation();

    if(edge.properties.length > 0) {
      return event.currentTarget.parentNode.classList.toggle('open');
    }

    return event.preventDefault();
  }

  componentDidMount() {
    document.addEventListener('click', this.handleOutsideClick, false);
  }

  componentWillUnmount() {
    document.removeEventListener('click', this.handleOutsideClick, false);
  }

}

export default NodeEdges;
