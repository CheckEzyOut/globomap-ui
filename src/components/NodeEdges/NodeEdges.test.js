import React from 'react';
import { shallow } from 'enzyme';
import { NodeEdges } from './NodeEdges';

describe("NodeEdges", function() {

  it('should render NodeEdges component', () => {
    const component = shallow(<NodeEdges />);
    expect(component).toMatchSnapshot();
  });

});
