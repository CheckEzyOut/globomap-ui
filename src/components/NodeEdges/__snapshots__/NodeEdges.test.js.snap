// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NodeEdges should render NodeEdges component 1`] = `
<div
  className="sub-node-edges"
/>
`;
